python/lsst/daf/butler_admin/script/update_datastore_checksums.py

```diff
@@ -0,0 +1,134 @@
+from typing import cast
+
+from lsst.daf.butler import Butler, DatasetRef
+from lsst.daf.butler.datastore import DatasetRefURIs
+from lsst.daf.butler.datastore.stored_file_info import StoredFileInfo
+from lsst.daf.butler.datastores.fileDatastore import FileDatastore
+from lsst.daf.butler.registry.interfaces import DatabaseInsertMode
+from lsst.resources import ResourcePath
+from lsst.utils.logging import getLogger
+
+_LOG = getLogger(__name__)
+
+
+def _compute_and_populate_checksum(
+    datastore: FileDatastore, uris: dict[DatasetRef, DatasetRefURIs], dataset_ref: DatasetRef
+) -> tuple[DatasetRef, StoredFileInfo] | None:
+    """Compute checksum and update the record's stored info if checksum
+    doesn't exist.
+
+    Parameters
+    ----------
+    datastore : `FileDatastore`
+        Datastore.
+    uris : `dict[DatasetRef, DatasetRefURIs]`
+        Dictionary containing the dataset reference to its URI.
+    dataset_ref : `DatasetRef`
+        Dataset reference.
+
+    Returns
+    -------
+    `Optional[tuple[DatasetRef, StoredFileInfo]]`
+        Dataset reference with its updated stored info.
+    """
+    stored_info = datastore.getStoredItemsInfo(dataset_ref)[0]
+
+    if stored_info.checksum:
+        return None
+    else:
+        primary_uri = uris[dataset_ref].primaryURI
+
+        if primary_uri is None:
+            raise ValueError(f"No URI found for dataset: {dataset_ref}")
+
+        resource_path = ResourcePath(primary_uri.path)
+
+        checksum = FileDatastore.computeChecksum(resource_path, algorithm="md5")
+
+        return dataset_ref, stored_info.update(checksum=checksum)
+
+
+def update_datastore_checksums(
+    repo: str,
+    dataset_type: Iterable[str],
+    collections: Iterable[str],
+    where: str,
+    find_first: bool,
+    limit: int,
+) -> None:
+    """Compute the checksum for the specified datasets in parallel
+    and update the file_datastore_records table.
+
+    Parameters
+    ----------
+    repo : `str`
+        URI of butler repository to update.
+    dataset_type : `Iterable[str]`
+        Names of dataset types.
+    collections : `Iterable[str]`
+        Names of collections.
+    where : `str`
+        Query string.
+    find_first : `bool`
+        Whether to find the first match or not.
+    limit : `int`
+        Limit the number of results to be returned. A value of 0 means
+        unlimited. A negative value is used to specify a cap where a warning
+        is issued if that cap is hit.
+    """
+    butler = Butler.from_config(repo, writeable=True)
+    datastore = cast(FileDatastore, butler._datastore)
+
+    for dt in dataset_type:
+        dataset_references = butler.query_datasets(
+            dt, collections, find_first=find_first, where=where, limit=limit
+        )
+
+        uris = datastore.getManyURIs(dataset_references)
+
+        prepare_checksum_task = partial(_compute_and_populate_checksum, datastore, uris)
+
+        with ThreadPoolExecutor() as executor:
```

Comment:
Have you tried ProcessPool? We find that for S3 datastores that process pool can be faster (but the main problem here is going to be downloading every file to temp storage anyhow).

Note that for threads the S3 downloads are going to complain a lot if you use more than 10 threads because that is the size of the `urllib3` connection pool by default. It might make more sense in the future for us to move the checksum code to `lsst.resources` and then add an `lsst.resources.ResourcePath.mchecksum()` so that we can keep all the worker pool code self-consistent.